src/appleseed.python/bindblenderprogressivetilecallback.cpp

```diff
@@ -0,0 +1,291 @@
+                        GL_RGBA,
+                        GL_FLOAT,
+                        m_buffer.data());
+                    glBindTexture(GL_TEXTURE_2D, 0);
+                    m_updated_buffer = false;
+                }
+
+                glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
+
+                glEnable(GL_TEXTURE_2D);
+                glBindTexture(GL_TEXTURE_2D, m_texture_id);
+                glBegin(GL_QUADS);
+                    glTexCoord2f(0.0f, 1.0f); glVertex3f(x    ,  y    , 0.0f);
+                    glTexCoord2f(1.0f, 1.0f); glVertex3f(x + w,  y    , 0.0f);
+                    glTexCoord2f(1.0f, 0.0f); glVertex3f(x + w,  y + h, 0.0f);
+                    glTexCoord2f(0.0f, 0.0f); glVertex3f(x    ,  y + h, 0.0f);
+                glEnd();
+                glBindTexture(GL_TEXTURE_2D, 0);
+                glDisable(GL_TEXTURE_2D);
+            }
+        }
+
+      private:
+        std::vector<float>  m_buffer;
+        size_t              m_buffer_width;
+        size_t              m_buffer_height;
+        bool                m_updated_buffer;
+
+        GLuint              m_texture_id;
+        size_t              m_texture_width;
+        size_t              m_texture_height;
+
+        bpy::object         m_request_redraw_callback;
+
+        void delete_texture()
+        {
+            if (m_texture_id != 0)
+            {
+                glDeleteTextures(1, &m_texture_id);
+
+                m_texture_id = 0;
+                m_texture_width = 0;
+                m_texture_height = 0;
+            }
+        }
+
+        void allocate_texture()
+        {
+            assert(m_texture_id == 0);
+
+            // We didn't receive any pixel yet.
+            if (m_buffer_width == 0 || m_buffer_height == 0)
+                return;
+
+            glGenTextures(1, &m_texture_id);
+
+            m_texture_width = m_buffer_width;
+            m_texture_height = m_buffer_height;
+
+            glBindTexture(GL_TEXTURE_2D, m_texture_id);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP);
+            glTexImage2D(
+                GL_TEXTURE_2D,
+                0,
+                GL_RGBA32F,
+                static_cast<GLsizei>(m_texture_width),
+                static_cast<GLsizei>(m_texture_height),
+                0,
+                GL_RGBA,
+                GL_FLOAT,
+                nullptr);
+            glBindTexture(GL_TEXTURE_2D, 0);
+        }
+
+        void copy_tile(const Tile& tile, const CanvasProperties& props, const size_t tile_x, const size_t tile_y)
+        {
+            const size_t x0 = tile_x * props.m_tile_width;
+            const size_t y0 = tile_y * props.m_tile_height;
+
+            for (size_t j = 0; j < props.m_tile_height; ++j)
```

Comment:
You need to use the _actual_ dimensions of the tile: border tiles may be smaller than others. Also, it'd be clearer to use `x` and `y` instead of `i` and `j` when iterating over pixels.

    for (size_t y = 0, h = tile.get_height(); y < h; ++y)